pkg/storage/chunk/objectclient/client.go

```diff
@@ -13,14 +14,30 @@ import (
 
 // KeyEncoder is used to encode chunk keys before writing/retrieving chunks
 // from the underlying ObjectClient
-type KeyEncoder func(string) string
+// Schema/Chunk are passed as arguments to allow this to improve over revisions
+type KeyEncoder func(schema chunk.SchemaConfig, chk chunk.Chunk) string
 
-// Base64Encoder is used to encode chunk keys in base64 before storing/retrieving
+// base64Encoder is used to encode chunk keys in base64 before storing/retrieving
 // them from the ObjectClient
-var Base64Encoder = func(key string) string {
+var base64Encoder = func(key string) string {
 	return base64.StdEncoding.EncodeToString([]byte(key))
 }
 
+var FSEncoder = func(schema chunk.SchemaConfig, chk chunk.Chunk) string {
+	// Filesystem encoder pre-v12 encodes the chunk as one base64 string.
+	// This has the downside of making them opaque and storing all chunks in a single
+	// directory, hurting performance at scale and discoverability.
+	// Post v12, we respect the directory structure imposed by chunk keys.
+	key := schema.ExternalKey(chk)
+	if schema.VersionForChunk(chk) > 11 {
+		split := strings.LastIndexByte(key, '/')
+		encodedTail := base64Encoder(key[split+1:])
+		return strings.Join([]string{key[:split], encodedTail}, "/")
```

Comment:
nit: should we use `filepath.Join()` here?

Reply:
We can't use `filepath` to build the keys when using `fsObjectClient` because it converts `/` to os specific separator using `filepath.FromSlash` and `filepath.ToSlash`. Basically, you would always use a `/` when using `fsObjectClient` , irrespective of os type .

If at all, we could use `path.Join` with always uses `/`, but it does not seem necessary since we are anyways working with a `/` above.

Reply:
> because it converts / to os specific separator using filepath.FromSlash and filepath.ToSlash.

That's why I thought we should use it, to have files in individual directories, also on windows.
With `strings.Join()` it would result in files all in the same root directory? But on unix system it would be organized in subfolders.

Reply:
Maybe I get something wrong here.

Reply:
What @chaudum said seems sound to me. The only time I can think where this would be problematic is mounting the same file system on different OS's (i.e. migrating on an NFS mounted Loki deployment), but I'm not sure that's something we can reasonably protect against.

Reply:
I just think it's weird to have different file/directory structure on Unix and Windows.